cmd/opm/root/add.go

```diff
@@ -0,0 +1,113 @@
+package root
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/operator-framework/operator-registry/pkg/action"
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"github.com/operator-framework/operator-registry/pkg/image/containerdregistry"
+	"github.com/operator-framework/operator-registry/pkg/lib/certs"
+)
+
+const (
+	retryInterval = time.Second * 5
+	timeout       = time.Minute * 1
+)
+
+type add struct {
+	logger     *logrus.Entry
+	configsDir string
+	debug      bool
+	caFile     string
+	pullTool   string
+	skipTLS    bool
+}
+
+func newAddCmd() *cobra.Command {
+	logger := logrus.New()
+	a := add{
+		logger:   logrus.NewEntry(logger),
+		pullTool: "none",
+	}
+	rootCmd := &cobra.Command{
+		Use:   "add <configs_path> <bundle_image1> <bundle_image2>........<bundle_imageN>",
+		Short: "add operator bundle/s to a catalog of packages",
+		Long:  `add operator bundles to a directory of configs representing packages in the catalog`,
+		Args:  cobra.MinimumNArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			a.configsDir = args[0]
+			if a.debug {
+				logger.SetLevel(logrus.DebugLevel)
+			}
+			return nil
+		},
+
+		RunE: a.addFunc,
+	}
+
+	rootCmd.Flags().BoolVar(&a.debug, "debug", false, "enable debug logging")
+	rootCmd.Flags().StringVarP(&a.caFile, "ca-file", "", "", "the root Certificates to use with this command")
+	rootCmd.Flags().BoolVar(&a.skipTLS, "skip-tls", false, "disable TLS verification")
+	return rootCmd
+}
+
+func (a *add) addFunc(cmd *cobra.Command, args []string) error {
+	rootCAs, err := certs.RootCAs(a.caFile)
+	if err != nil {
+		return fmt.Errorf("failed to get RootCAs: %v", err)
+	}
+	reg, err := containerdregistry.NewRegistry(containerdregistry.SkipTLS(a.skipTLS), containerdregistry.WithLog(a.logger), containerdregistry.WithRootCAs(rootCAs))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := reg.Destroy(); err != nil {
+			a.logger.Errorf("error destroying local cache: %v", err)
+		}
+	}()
+	bundles := []action.InputBundle{}
+	for _, ref := range args[1:] {
+		simpleRef := image.SimpleReference(ref)
+		tmpDir, err := ioutil.TempDir("./", "bundle_tmp")
+		if err != nil {
+			return fmt.Errorf("error creating temp directory to unpack bundle image %q in:%v", simpleRef.String(), err)
+		}
+		defer func() {
+			a.logger.Infof("Removing temp directory %q bundle was unpacked in", tmpDir)
+			if err := os.RemoveAll(tmpDir); err != nil {
+				a.logger.Errorf("error removing temp directory %q bundle was unpacked in: %v", tmpDir, err)
+			}
+		}()
```

Comment:
Using defer in a for loop means that a bunch of deferred functions will be added to the defer stack and they only run when this entire function exits.

It might be better to encapsulate the bundle extract logic into its own function so that we clean up after ourselves after each bundle instead of after all bundles.

Reply:
>Using defer in a for loop means that a bunch of deferred functions will be added to the defer stack and they only run when this entire function exits.

Isn't that what we want? 

1. Function begins 
2. Extract all the bundles into tmp folders
3. Call AddToConfig(request{bundles: tmpFolders})
4. Clean up all the bundles

Is the concern here that the defer stack will morph into something we cannot predict? If that's the case we'll have to say we cannot depend of defer, and instead re-design AddToConfig to only ever add one bundle at a time (which imho I was happy to avoid by leaning on the defer stack)

Reply:
I see. I had in my mind that we would internally just work on one image at a time. So something like:
1. Load Declcfg, convert to model
2. For each image
   - Pull/extract it
   - Add it to to the model
   - Cleanup the extracted dir
3. Convert model to declcfg, write declcfg

Reply:
My big concern is just the high-water mark of disk space used.

Reply:
I see your point. But here's the counter argument: 

> we would internally just work on one image at a time

that's not the contract you're signing when you're passing multiple bundle images in the args. You're saying "adding all of these bundles at once". If you are about to add 100s of bundles, it's better to chop that job up into pieces. 

The reason I say this is because this flow: 

```
1. Load Declcfg, convert to model
2. For each image
      Pull/extract it
      Add it to to the model
      Cleanup the extracted dir
6. Convert model to declcfg, write declcfg
```
translates to 
Load declcfg, convert to model 
For each image, AddBundleToModel (instead of AddToConifg) 
And that doesn't seem like the right way to go since that assumes that any and all consumers of BundleAdder must know what the model is, or at the very least must deal with the model before passing it on to BundleAdder, whereas the model is supposed to be an internal model?

Reply:
@joelanford changed the design of `BundleAdder` to unpack and remove one bundle at a time.